### PR TITLE
 e2e: test: introduce workarounds

### PR DESCRIPTION
**Pull request by ffromani**
formalize workarounds we need to add to unblock CI.
We should strive to keep this list short, but a sad
reality is that we need workarounds from time to time.

### CODE REVIEW

**Comment by ffromani** [comment]
counterproof this is still required: https://github.com/openshift-kni/numaresources-operator/pull/1098#issuecomment-2520551032

File: internal/workarounds/prom.go
Comment (shajmakh) [inline internal/workarounds/prom.go:76]: do you want the loop to be aborted if you find another error? if not you need to return nil instead
and regardless why to return true? the condition isn't satisfied

File: internal/workarounds/prom.go
Comment (shajmakh) [inline internal/workarounds/prom.go:81]: same here, I believe this should be return false, nil and log out the error as warning unless I'm missing the point of retrying

**Review by shajmakh — COMMENTED** [review]
cool implementation!

File: internal/workarounds/prom.go
Comment (ffromani) [reply internal/workarounds/prom.go:76]: yes, this failure is critical and needs to make the process abort

File: internal/workarounds/prom.go
Comment (ffromani) [reply internal/workarounds/prom.go:81]: again failure at this stage is critical, to put it differently, the only non-critical failure is on update, because tolerating failure on update is needed for optimistic concurrency

File: internal/workarounds/prom.go
Comment (Tal-or) [inline internal/workarounds/prom.go:55]: This issue is closed, then what is the indication for us that this workaround is not needed anymore?

File: internal/workarounds/prom.go
Comment (ffromani) [reply internal/workarounds/prom.go:59]: good point. I'll re-upload with an explanation comment. I want to have another full run to demonstrate the W/A is fixing (or at least non-hurting) and then I'll re-upload

**Review by Tal-or — APPROVED** [review]
I'll be honest, I think we're going too far with this one.
There are too many questions here:
1. How we're going to decide when an external issue deserves a workaround? every time CI failed? which lanes? 
 
2. When issues like these reoccur (and they will)  are we going to spent hours and days understanding the issue and workaround it? why overriding the issue and testing locally is not enough? 

3. How do we know when issue get solved and we don't need the workaround anymore? what if the workaround will break?

I can understand the rational here, but this is a rabbit hole of maintenance and extra work the IMVHO is too costly.

**Comment by Tal-or** [comment]
/hold
Got comments from other reviewers

**Comment by ffromani** [comment]
> I'll be honest, I think we're going too far with this one. There are too many questions here:

thing is: you are right, full stop, no "if" "but" or "however". I don't like this idea any more than you do (even though, I must say, the impl turned out quite nice!). But we're past time and we have urgent matters to cover. So this is a necessary evil.

>      1. How we're going to decide when an external issue deserves a workaround? every time CI failed? which lanes?

Never. We are out of options at this particular point in time, and I wanted some structure to convey the key details. I maybe went overboard so I'm fine with narrowing down, even though that won't solve the fundamental issue that we should not have any code at all to workaround issues.

>     2. When issues like these reoccur (and they will)  are we going to spent hours and days understanding the issue and workaround it? why overriding the issue and testing locally is not enough?

The first part yes, unfortunately: we will need to painfully spend hours and days to be able to report the core issue to the relevant team (like we are doing here). Workarounding is a necessary evil. Testing locally is not enough: things degrade fast and hard without any form of automated CI, and our CI is already bare bones (serial suite can't run here :\ )

>     3. How do we know when issue get solved and we don't need the workaround anymore? what if the workaround will break?

We will need to keep going the right route and file issues to get the root cause addressed. In parallel we need to unblock ourselves as we are out of time. The workaround can (and will) very much break, and at that point we will be back at square one.

> I can understand the rational here, but this is a rabbit hole of maintenance and extra work that IMVHO is too costly.

While I do agree, the problem is that *just* waiting (vs workarounding and waiting for the fix in parallel) is even costlier.

> EDIT: I approved the changes because I understand we're on the edge here.

Yes, it's a necessary evil. I'm not happier than you are, likely less happy.

File: internal/workarounds/prom.go
Comment (ffromani) [reply internal/workarounds/prom.go:55]: the u/s issue is fixed, yet we are hitting the same issue in our CI, and I don't have yet a better bug to track.
Once this fixes and is merged, I'll post a revert to track the real fix.

**Comment by shajmakh** [comment]
The implementation looks good and I understand the raised concerns, I am willing to maintain if needed
/lgtm 
feel free to unhold as you see fit

**Comment by ffromani** [comment]
turns out the actual culprit can be crun 1.18.z (vs previous 1.17) but the investigation is still ongoing and we don't have a repro (yet alone a fix) yet.

I'm taking the responsability to unhold, because we need to move forward with other issues and tasks

**Comment by ffromani** [comment]
https://issues.redhat.com/browse/OCPBUGS-45394

**Pull request merged**
